drivers/fxos8700/fxos8700.c

```diff
@@ -0,0 +1,205 @@
+        return FXOS8700_NOBUS;
+    }
+    if (config != FXOS8700_WHO_AM_I_VAL) {
+        i2c_release(dev->p.i2c);
+        DEBUG("[fxos8700] WHOAMI is wrong (%2x)\n", config);
+        return FXOS8700_NOBUS;
+    }
+
+    /* Configure the ODR to maximum (400Hz in hybrid mode) */
+    config = 0x00;
+    if (fxos8700_write_regs(dev, FXOS8700_REG_CTRL_REG1, &config, 1) != FXOS8700_OK) {
+        return FXOS8700_NOBUS;
+    }
+    /* Activate hybrid mode */
+    config = HYBRID_MODE;
+    if (fxos8700_write_regs(dev, FXOS8700_REG_M_CTRL_REG1, &config, 1) != FXOS8700_OK) {
+        return FXOS8700_NOBUS;
+    }
+    /* Set burst read mode (accel + magnet together) */
+    config = 0x20;
+    if (fxos8700_write_regs(dev, FXOS8700_REG_M_CTRL_REG2, &config, 1) != FXOS8700_OK) {
+        return FXOS8700_NOBUS;
+    }
+
+    /* initial read for caching operation */
+    if (fxos8700_read(dev, &acc_cached, &mag_cached) != FXOS8700_OK) {
+        return FXOS8700_BUSERR;
+    }
+    last_read_time = xtimer_now_usec();
+
+    return FXOS8700_OK;
+}
+
+
+int fxos8700_set_active(fxos8700_t* dev)
+{
+    uint8_t config = 0x01;
+    if (fxos8700_write_regs(dev, FXOS8700_REG_CTRL_REG1, &config, 1) != FXOS8700_OK) {
+        return FXOS8700_BUSERR;
+    }
+    return FXOS8700_OK;
+}
+
+int fxos8700_set_idle(fxos8700_t* dev)
+{
+    uint8_t config = 0x00;
+    if (fxos8700_write_regs(dev, FXOS8700_REG_CTRL_REG1, &config, 1) != FXOS8700_OK) {
+        return FXOS8700_BUSERR;
+    }
+    return FXOS8700_OK;
+}
+
+int fxos8700_read(const fxos8700_t* dev, fxos8700_measurement_t* acc, fxos8700_measurement_t* mag)
+{
+    uint8_t data[12];
+    uint8_t ready = 0;
+
+    if (fxos8700_set_active(dev)) {
+        return FXOS8700_BUSERR;
+    }
+
+    while(!(ready & 0x08)) {
+        fxos8700_read_regs(dev, FXOS8700_REG__STATUS, &ready, 1);
+    }
+    while(!(ready & 0x08)) {
+        fxos8700_read_regs(dev, FXOS8700_REG_M_DR_STATUS, &ready, 1);
+    }
+
+    /* Read all data at once */
+    if (fxos8700_read_regs(dev, FXOS8700_REG_OUT_X_MSB, &data[0], 12)) {
+        return FXOS8700_BUSERR;
+    }
+
+    if (fxos8700_set_idle(dev)) {
+        return FXOS8700_BUSERR;
+    }
+
+    /* Read accelerometer */
+    if (acc) {
+        int32_t acc_raw_x = (int32_t) ((data[0]<<6) | (data[1]>>2));
+        int32_t acc_raw_y = (int32_t) ((data[2]<<6) | (data[3]>>2));
+        int32_t acc_raw_z = (int32_t) ((data[4]<<6) | (data[5]>>2));
+        acc->x = (int16_t)(acc_raw_x*244/1000);
+        acc->y = (int16_t)(acc_raw_y*244/1000);
+        acc->z = (int16_t)(acc_raw_z*244/1000);
+    }
+    /* Read magnetometer */
+    if (mag) {
+        mag->x = (int16_t) ((data[6] <<8) | data[7]);
+        mag->y = (int16_t) ((data[8] <<8) | data[9]);
+        mag->z = (int16_t) ((data[10]<<8) | data[11]);
+    }
+    return FXOS8700_OK;
+}
+
+int fxos8700_read_cached(const fxos8700_t *dev, fxos8700_measurement_t* acc, fxos8700_measurement_t* mag)
+{
+    uint32_t now = xtimer_now_usec();
```

Comment:
Optional: It may be less computationally expensive to use xtimer_now and convert the renew interval to ticks instead, since you can precompute that us to tick conversion at compilation time rather than do a conversion from ticks to us in every read. See this as an optional comment, and maybe just remember for future drivers, I assume this function will be called somewhat infrequently since the renew interval is set to 1 second.